### DISCUSSION TITLE
Provide actix-web v4 doc or migration guide

### DISCUSSION THREAD

**Discussion by sessionboy** [actix/actix-web#2170] [General]
I am starting a new actix project, but the v4 version seems to have major changes, so in order to avoid possible migration risks, I want to start with v4.

Hope team can provide v4 documentation or migration guide.

**Comment by robjtede** [comment]
Providing a migration guide is the plan. Before the stable release we'll collate the most impactful changes from the changelog and expand on them in MIGRATION.md.

**Reply by sessionboy** [reply]
How long is it estimated? I plan to use v4 in a new project, but there is no related documentation and I don’t know how to start.

**Reply by fakeshadow** [reply]
From what I see all docs are built properly. Is this not work for you? https://docs.rs/actix-web/4.0.0-beta.6/actix_web/index.html

Most of the break changes can be found in CHANGES.md and you would want to couple with docs.rs to figure out the changes.

betas contains incremental break changes so provide a migration guide early could be more work to do and further delay the stable release.

**Reply by sessionboy** [reply]
@fakeshadow  Thanks. 
> From what I see all docs are built properly. Is this not work for you? https://docs.rs/actix-web/4.0.0-beta.6/actix_web/index.html.    

I am familiar with v3, but not familiar with the changes of v4. Without documentation, I can only find and debug without a clue, which is very inefficient.

> betas contains incremental break changes so provide a migration guide early could be more work to do and further delay the stable release.    

Yeah, Maybe you don’t need to write detailed documents, but you can list some simple examples that you use in your tests. 

Thank you for your efforts. 🍻 Looking forward to the early release of v4.

**Reply by fakeshadow** [reply]
> For example, `web::HttpResponse` in v3 has become `web::HttpResponse<B=Body>` in v4, its usage has changed a lot. In the absence of examples, I still need to spend time exploring how to use it. I don't know how many such changes remain, everything is explored in the unknown.

I didn't make changes to it so I'm not sure what's the case on `Response` types either. I believe @robjtede could give some guide on it.

> Yeah, Maybe you don’t need to write detailed documents, but you can list some simple examples that you use in your tests.

Yea I would like to do add examples on new features and changes but still get some other issues not figured out for next release(Possibly).

> Thank you for your efforts. 🍻 Looking forward to the early release of v4.

You are welcome. Sorry for keep you waiting we want to make it right and ready and possibly keep v4 stable for a much longer time than v3.

**Reply by sessionboy** [reply]
> I didn't make changes to it so I'm not sure what's the case on Response types either.

I don't know, but I got a error prompt in v4: 
```rust
impl actix_web::ResponseError for ServerError {
  fn error_response(&self) -> web::HttpResponse {
    // todo
  }
 
  // pub struct Response<B>
  // An HTTP response.

  // missing generics for struct `actix_web::BaseHttpResponse`

  // expected 1 type argument

  // help: use angle brackets to add missing type argument: `<B>`
}
```

After changing to the following, it runs normally:
```rust
use actix_web::{ web, http };
use actix_http::{body::Body, http::HeaderValue, http::StatusCode};
...
impl actix_web::ResponseError for ServerError {
  fn error_response(&self) -> web::HttpResponse<Body> {  // change
    // todo
  }
}
```
There are other changes, and I am still exploring it.

> Yea I would like to do add examples on new features and changes but still get some other issues not figured out for next release(Possibly).

I am looking forward.  👏
